plugins/modules/foreman_operating_system.py

```diff
@@ -139,47 +166,38 @@
 
 RETURN = ''' # '''
 
-try:
-    from ansible.module_utils.ansible_nailgun_cement import (
-        find_entities,
-        find_entities_by_name,
-        find_operating_system_by_title,
-        naildown_entity,
-        naildown_entity_state,
-        sanitize_entity_dict,
-        OperatingSystemParameter,
-    )
 
-    from nailgun.entities import (
-        Architecture,
-        Media,
-        OperatingSystem,
-        PartitionTable,
-        ProvisioningTemplate,
-    )
-except ImportError:
-    pass
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule, parameter_value_to_str
 
-from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
 
 # This is the only true source for names (and conversions thereof)
-name_map = {
-    'name': 'name',
-    'description': 'description',
-    'family': 'family',
-    'major': 'major',
-    'minor': 'minor',
-    'release_name': 'release_name',
-    'architectures': 'architecture',
-    'media': 'medium',
-    'ptables': 'ptable',
-    'provisioning_templates': 'provisioning_template',
-    'password_hash': 'password_hash',
+entity_spec = {
+    'id': {},
+    'name': {},
+    'description': {},
+    'family': {},
+    'major': {},
+    'minor': {},
+    'release_name': {},
+    'architectures': {'type': 'entity_list', 'flat_name': 'architecture_ids'},
+    'media': {'type': 'entity_list', 'flat_name': 'medium_ids'},
+    'ptables': {'type': 'entity_list', 'flat_name': 'ptable_ids'},
+    'provisioning_templates': {'type': 'entity_list', 'flat_name': 'provisioning_template_ids'},
+    'password_hash': {},
+}
```

Comment:
@evgeni After having real trouble updating nested entities, i started to elaborate on your foreman_spec idea. I came up with an entity_spec and a rewrite of the complete ensure mechanism (this time tailored for dict:dict instead of dict:nailgun).
One nice but nonobvious side effect is that i could reduce the doubled translation via name_map and flattening into a single step by applying the entity_spec.
The _new_ update mechanism does not yet know the specialities like `force_update`, `check_missing` ... I think, we can add those on the go, and specify their need in the entity_spec. I hope, we keep the update function clean on the way.

Reply:
Please consider adding organization_id here or on global for all resources

Reply:
True, sorry, confused it with other resources - https://github.com/theforeman/foreman-ansible-modules/issues/365